Sources/PenguinStructures/Concatenation.swift

```diff
@@ -0,0 +1,203 @@
+      return first.distance(from: start, to: first.endIndex) + second.distance(from: second.startIndex, to: end)
+    case (.b(let start), .a(let end)):
+      return second.distance(from: start, to: second.startIndex) + first.distance(from: first.endIndex, to: end)
+    case (.b(let start), .b(let end)):
+      return second.distance(from: start, to: end)
+    }
+  }
+}
+
+extension Concatenation: BidirectionalCollection
+where First: BidirectionalCollection, Second: BidirectionalCollection {
+  /// Returns the next position before `i`.
+  @inlinable
+  public func index(before i: Index) -> Index {
+    switch i.position {
+    case .a(let index): return Index(first: first.index(before: index))
+    case .b(let index):
+      if index == second.startIndex {
+        return Index(first: first.index(before: first.endIndex))
+      }
+      return Index(second: second.index(before: index))
+    }
+  }
+}
+
+extension Concatenation: RandomAccessCollection
+  where First: RandomAccessCollection, Second: RandomAccessCollection
+{
+  @inlinable
+  public func index(_ i: Index, offsetBy n: Int) -> Index {
+    if n == 0 { return i }
+    if n < 0 { return offsetBackward(i, by: n) }
+    return offsetForward(i, by: n)
+  }
+
+  @usableFromInline
+  func offsetForward(_ i: Index, by n: Int) -> Index {
+    switch i.position {
+    case .a(let index):
+      let d = first.distance(from: index, to: first.endIndex)
+      if n < d {
+        return Index(first: first.index(index, offsetBy: n))
+      } else {
+        return Index(second: second.index(second.startIndex, offsetBy: n - d))
+      }
+    case .b(let index):
+      return Index(second: second.index(index, offsetBy: n))
+    }
+  }
+
+  @usableFromInline
+  func offsetBackward(_ i: Index, by n: Int) -> Index {
+    switch i.position {
+    case .a(let index):
+      return Index(first: first.index(index, offsetBy: n))
+    case .b(let index):
+      let d = second.distance(from: second.startIndex, to: index)
+      if -n <= d {
+        return Index(second: second.index(index, offsetBy: n))
+      } else {
+        return Index(first: first.index(first.endIndex, offsetBy: n + d))
+      }
+    }
+  }
+}
+
+extension Concatenation: MutableCollection
+  where First: MutableCollection, Second: MutableCollection
+{
+  /// Accesses the element at `i`.
+  @inlinable
+  public subscript(i: Index) -> Element {
+    get {
+      switch i.position {
+      case .a(let index): return first[index]
+      case .b(let index): return second[index]
+      }
+    }
+    set {
+      switch i.position {
+      case .a(let index): first[index] = newValue
+      case .b(let index): second[index] = newValue
+      }
+    }
+  }
+}
+
+// TODO: Add MutableCollection conformance.
+
+extension Collection {
+  /// Returns a new collection where all the elements of `self` appear before all the elements of
+  /// `other`.
+  @inlinable
+  public func joined<Other: Collection>(to other: Other) -> Concatenation<Self, Other>
```

Comment:
Is joined the right name here? Or should we rename the type?

Reply:
It's a tough call.  Maybe `concatenated(to:)`.  It was the `with:` that bothered me most as it appeared to disclaim any description of the ordering, and perhaps I took it too far.

https://developer.apple.com/documentation/swift/array/1690077-joined

is significantly different in that it includes a separator and also it produces a `JoinedCollection`, so there's precedent for naming consistency here.

Reply:
Changed.